tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,727 @@
+//! # Implementation Details
+//!
+//! The semaphore is implemented using an intrusive linked list of waiters. An
+//! atomic counter tracks the number of available permits. If the semaphore does
+//! not contain the required number of permits, the task attempting to acquire
+//! permits places its waker at the end of a queue. When new permits are made
+//! available (such as by releasing an initial acquisition), they are assigned
+//! to the task at the front of the queue, waking that task if its requested
+//! number of permits is met. 
+//!
+//! Because waiters are enqueued at the back of the linked list and dequeued
+//! from the front, the semaphore is fair. Tasks trying to acquire large numbers
+//! of permits at a time will always be woken eventually, even if many other
+//! tasks are acquiring smaller numbers of permits. This means that in a
+//! use-case like tokio's read-write lock, writers will not be starved by
+//! readers.
+//!
+//! The linked list is guarded by a mutex, which must be acquired before
+//! enqueueing or dequeueing a task. However, some operations are always
+//! wait-free.
+use crate::loom::{
+    cell::CausalCell,
+    sync::{atomic::AtomicUsize, Mutex, MutexGuard},
+};
+use crate::util::linked_list::{self, LinkedList};
+
+use std::{
```

Comment:
Stylistically, we are avoiding this import style in the tokio code base. i.e. no nested `{ .. }`, instead different lines. IMO, if we want to change this, it should be done explicitly and w/ a PR to update existing code.

Reply:
okay, will change that --- i think rustfmt defaulted to this?